### PR TITLE
Quantstamp audit base 2

### PR DESCRIPTION
**Pull request by akshay111meher**

### CODE REVIEW

File: contracts/CreditLine/CreditLine.sol [hunk 1]
```diff
@@ -569,15 +582,21 @@ contract CreditLine is CreditLineStorage, ReentrancyGuard {
...
             (creditLineUsage[creditLineHash].interestAccruedTillPrincipalUpdate).add(_interestSincePrincipalUpdate);
         uint256 _totalDebt = _totalInterestAccrued.add(creditLineUsage[creditLineHash].principal);
         uint256 _totalRepaidNow = creditLineUsage[creditLineHash].totalInterestRepaid.add(repayAmount);
-        require(_totalDebt >= repayAmount, 'CreditLine: Repay amount is greater than debt.');
+
+        bool _totalRemainingIsRepaid = false;
+
+        if (repayAmount > _totalDebt) {
+            _totalRemainingIsRepaid = true;
+            repayAmount = _totalDebt;
```
Comment (prateekreddy) [inline contracts/CreditLine/CreditLine.sol:598]: repay() function assumes that repayAmount is equal to msg.value in case of ether is the asset. So repay will fail if repayAmount > _totalDebt

File: contracts/CreditLine/CreditLine.sol [hunk 2]
```diff
@@ -592,7 +611,12 @@ contract CreditLine is CreditLineStorage, ReentrancyGuard {
         if (creditLineUsage[creditLineHash].principal == 0) {
             _resetCreditLine(creditLineHash);
         }
-        emit PartialCreditLineRepaid(creditLineHash, repayAmount);
+
+        if (_totalRemainingIsRepaid) {
```
Comment (prateekreddy) [inline contracts/CreditLine/CreditLine.sol:615]: Make appropriate changes as per changes to line 598

File: contracts/CreditLine/CreditLine.sol [hunk 3]
```diff
@@ -687,21 +715,14 @@ contract CreditLine is CreditLineStorage, ReentrancyGuard {
             uint256 _tokensToTransfer = _tokenInStrategy;
             if (_activeAmount.add(_tokenInStrategy) > _amountInTokens) {
                 _tokensToTransfer = _amountInTokens.sub(_activeAmount);
-
                 liquidityShares = liquidityShares.mul(_tokensToTransfer).div(_tokenInStrategy);
-            }
-            _activeAmount = _activeAmount.add(_tokensToTransfer);
+            } // 0
+            _activeAmount = _activeAmount.add(_tokensToTransfer); //64753966145743327485
```
Comment (prateekreddy) [inline contracts/CreditLine/CreditLine.sol]: please remove these numbers

File: contracts/Pool/Pool.sol [hunk 4]
```diff
@@ -1,6 +1,7 @@
 // SPDX-License-Identifier: MIT
 pragma solidity 0.7.0;
 
+import '@openzeppelin/contracts/token/ERC20/ERC20.sol';
```
Comment (prateekreddy) [inline contracts/Pool/Pool.sol:4]: why was this added if SafeERC20 is already there ?

File: contracts/Pool/Pool.sol [hunk 5]
```diff
@@ -683,16 +680,16 @@ contract Pool is Initializable, IPool, ReentrancyGuard {
...
             correspondingBorrowTokens(_collateralTokens, _poolFactory, IPoolFactory(_poolFactory).liquidatorRewardFraction());
 
         delete poolVars.extraLiquidityShares;
         delete poolVars.baseLiquidityShares;
-        
+
         _deposit(_fromSavingsAccount, false, _borrowAsset, _poolBorrowTokens, address(0), msg.sender, address(this));
         _withdraw(_toSavingsAccount, _recieveLiquidityShare, _collateralAsset, _poolSavingsStrategy, _collateralTokens);
-
+        delete poolVars.extraLiquidityShares;
```
Comment (prateekreddy) [inline contracts/Pool/Pool.sol]: These are already deleted in line 686 & 687

File: contracts/Repayments/Repayments.sol [hunk 6]
```diff
@@ -305,29 +305,21 @@ contract Repayments is Initializable, RepaymentStorage, IRepayment, ReentrancyGu
...
         if (_asset == address(0)) {
             require(_amountRequired <= msg.value, 'Repayments::repayAmount amount does not match message value.');
             (bool success, ) = payable(address(_poolID)).call{value: _amountRequired}('');
             require(success, 'Transfer failed');
         } else {
             IERC20(_asset).safeTransferFrom(msg.sender, _poolID, _amountRequired);
         }
-
-        if (_asset == address(0)) {
-            if (msg.value > _amountRequired) {
```
Comment (prateekreddy) [inline contracts/Repayments/Repayments.sol:324]: This is necessary to return extra ether back to the user. Why was this removed ?

File: contracts/SavingsAccount/SavingsAccountUtil.sol [hunk 7]
```diff
@@ -100,18 +101,22 @@ library SavingsAccountUtil {
...
+            require(msg.value >= _amount, 'ethers provided should be greater than _amount');
+
             if (_to != address(this)) {
                 (bool success, ) = payable(_to).call{value: _amount}('');
                 require(success, 'Transfer failed');
             }
             if (msg.value >= _amount) {
-                (bool success, ) = payable(address(msg.sender)).call{value: msg.value - _amount}('');
-                require(success, 'Transfer failed');
+                if (msg.value - _amount != 0) {
```
Comment (prateekreddy) [inline contracts/SavingsAccount/SavingsAccountUtil.sol:115]: Can we just remove = in the above if rather than use this if condition ?

File: contracts/yield/YearnYield.sol [hunk 8]
```diff
@@ -182,8 +182,9 @@ contract YearnYield is IYield, Initializable, OwnableUpgradeable, ReentrancyGuar
         uint256 initialAssetBalance = IERC20(asset).balanceOf(address(this));
 
         //withdraw collateral from vault
+        // IyVault(vault).withdraw(amount.mul(3).div(10));
         IyVault(vault).withdraw(amount);
-
+        // 24709119562098501221
```
Comment (prateekreddy) [inline contracts/yield/YearnYield.sol:187]: Can we remove the commented amounts as well as the above commented code for withdraw

File: utils/constants-rahul.ts [hunk 9]
```diff
@@ -0,0 +1,143 @@
+import { BigNumber } from '@ethersproject/providers/node_modules/@ethersproject/bignumber';
```
Comment (prateekreddy) [inline utils/constants-rahul.ts:1]: Why does this exists seperately ? Let's use a single constants file

File: utils/constants-Additions.ts [hunk 10]
```diff
@@ -0,0 +1,119 @@
+import { BigNumber } from '@ethersproject/bignumber';
```
Comment (prateekreddy) [inline utils/constants-Additions.ts:1]: Why does this exists seperately ? Let's use a single constants file

**Review by prateekreddy — CHANGES_REQUESTED** [review]

File: contracts/CreditLine/CreditLine.sol [hunk 11]
```diff
@@ -509,19 +507,34 @@ contract CreditLine is CreditLineStorage, ReentrancyGuard {
...
+
+        // @ TO-DO
+        uint256 _tokenDiffBalance;
+        if (_borrowAsset != address(0)) {
+            uint256 _balanceBefore = IERC20(_borrowAsset).balanceOf(address(this));
+            _withdrawBorrowAmount(_borrowAsset, borrowAmount, _lender);
+            uint256 _balanceAfter = IERC20(_borrowAsset).balanceOf(address(this));
+            _tokenDiffBalance = _balanceAfter.sub(_balanceBefore);
+        } else {
+            _withdrawBorrowAmount(_borrowAsset, borrowAmount, _lender);
```
[the remark this answers is not in the document]
Comment (akshay111meher) [reply contracts/CreditLine/CreditLine.sol:520]: Cleared!

File: contracts/CreditLine/CreditLine.sol [hunk 12]
```diff
@@ -509,19 +507,34 @@ contract CreditLine is CreditLineStorage, ReentrancyGuard {
...
             require(feeSuccess, 'Transfer fail');
-            (bool success, ) = msg.sender.call{value: borrowAmount}('');
+            (bool success, ) = msg.sender.call{value: _tokenDiffBalance}('');
             require(success, 'Transfer fail');
+            emit BorrowedFromCreditLine(_tokenDiffBalance, creditLineHash);
         } else {
             IERC20(_borrowAsset).safeTransfer(protocolFeeCollector, _protocolFee);
-            IERC20(_borrowAsset).safeTransfer(msg.sender, borrowAmount);
+            IERC20(_borrowAsset).safeTransfer(msg.sender, _tokenDiffBalance);
+            emit BorrowedFromCreditLine(_tokenDiffBalance, creditLineHash);
```
[the remark this answers is not in the document]
Comment (akshay111meher) [reply contracts/CreditLine/CreditLine.sol:536]: cleared!.

File: contracts/SavingsAccount/SavingsAccountUtil.sol [hunk 13]
```diff
@@ -112,11 +112,8 @@ library SavingsAccountUtil {
                 require(success, 'Transfer failed');
             }
             if (msg.value >= _amount) {
```
Comment (prateekreddy) [inline contracts/SavingsAccount/SavingsAccountUtil.sol]: Remove the equality so that we don't do 0 value transfers

**Review by prateekreddy — CHANGES_REQUESTED** [review]

**Review by prateekreddy — APPROVED** [review]
